chain/blocksync/blocksync_client.go

```diff
@@ -374,13 +399,26 @@ func (bpt *bsPeerTracker) prefSortedPeers() []peer.ID {
 		out = append(out, p)
 	}
 
+	// sort by 'expected cost' of requesting data from that peer
+	// additionally handle edge cases where not enough data is available
 	sort.Slice(out, func(i, j int) bool {
 		pi := bpt.peers[out[i]]
 		pj := bpt.peers[out[j]]
+
+		if pj.successes > 0 && pi.successes > 0 {
+			failRateI := float64(pi.failures) / float64(pi.failures+pi.successes)
+			costI := float64(pi.averageTime) * failRateI
```

Comment:
I would change it to `float64(pi.averageTime) * (1 + failRateI)`.
This way we will pick the best peer among peers that never failed us.